readline/buffer.go

```diff
@@ -0,0 +1,356 @@
+package readline
+
+import (
+	"fmt"
+
+	"github.com/emirpasic/gods/lists/arraylist"
+	"golang.org/x/term"
+)
+
+type Buffer struct {
+	Pos       int
+	Buf       *arraylist.List
+	Prompt    *Prompt
+	LineWidth int
+	Width     int
+	Height    int
+}
+
+func NewBuffer(prompt *Prompt) (*Buffer, error) {
+	width, height, err := term.GetSize(0)
+	if err != nil {
+		fmt.Println("Error getting size:", err)
+		return nil, err
+	}
+
+	lwidth := width - len(prompt.Prompt)
+	if prompt.UseAlt {
+		lwidth = width - len(prompt.AltPrompt)
+	}
+
+	b := &Buffer{
+		Pos:       0,
+		Buf:       arraylist.New(),
+		Prompt:    prompt,
+		Width:     width,
+		Height:    height,
+		LineWidth: lwidth,
+	}
+
+	return b, nil
+}
+
+func (b *Buffer) MoveLeft() {
+	if b.Pos > 0 {
+		if b.Pos%b.LineWidth == 0 {
+			fmt.Printf(CursorUp + CursorBOL + fmt.Sprintf(CursorRightN, b.Width))
```

Comment:
nit: this doesn't really need fmt.Printf.

I think `Cursor(Up|Down|Left|Right)N` might be better suited as functions rather than constants.

e.g.

```go
var cursorUp = cursorUpN(1)

func cursorUpN(n int) string {
  return fmt.Sprintf("\033[%dA", n)
}
```

This becomes `fmt.Print(cursorUp + CursorBOL + cursorUpN(b.Width))`

Reply:
I ended up implementing cursorUp/Down/Left/RightN.